shared/common-adapters/search-filter.js

```diff
@@ -0,0 +1,317 @@
+// @flow
+import * as React from 'react'
+import Box, {Box2} from './box'
+import ClickableBox from './clickable-box'
+import NewInput from './new-input'
+import PlainInput from './plain-input'
+import Text from './text'
+import ProgressIndicator from './progress-indicator'
+import Icon, {type IconType} from './icon'
+import * as Styles from '../styles'
+import * as Platforms from '../constants/platform'
+
+const KeyHandler = (Styles.isMobile ? c => c : require('../util/key-handler.desktop').default)(() => null)
+
+const Kb = {
+  Box,
+  Box2,
+  ClickableBox,
+  Icon,
+  NewInput,
+  PlainInput,
+  ProgressIndicator,
+  Text,
+}
+
+type Props = {|
+  icon?: ?IconType,
+  focusOnMount?: ?boolean,
+  negative?: ?boolean,
+  onChange: (text: string) => void,
+  placeholderText: string,
+  style?: ?Styles.StylesCrossPlatform,
+  type: 'Small' | 'Full-width' | 'Mobile',
+  waiting?: ?boolean,
+
+  onBlur?: ?() => void,
+  // If onClick is provided, this component won't focus on click. User is
+  // expected to handle actual filter/search in a separate component, perhaps
+  // in a popup.
+  onClick?: ?() => void,
+  // following props are ignored when onClick is provided
+  hotkey?: ?('f' | 'k'), // desktop only
+  onKeyDown?: (event: SyntheticKeyboardEvent<>, isComposingIME: boolean) => void,
+  onKeyUp?: (event: SyntheticKeyboardEvent<>, isComposingIME: boolean) => void,
+|}
+
+type State = {|
+  focused: boolean,
+  hover: boolean,
+  text: string,
+|}
+
+class SearchFilter extends React.PureComponent<Props, State> {
+  state = {
+    focused: false,
+    hover: false,
+    text: '',
+  }
+
+  _input = React.createRef()
+  _onBlur = () => {
+    this.setState({focused: false})
+    this.props.onBlur && this.props.onBlur()
+  }
+  _onFocus = () => {
+    this.setState({focused: true})
+  }
+
+  _focus = () => {
+    if (this.state.focused) {
+      // Can't just make a null for Kb.ClickableBox as that'd cause
```

Comment:
I don't understand this comment.

Reply:
When you pass in a `null` as `onClick` to `Kb.ClickableBox`, it becomes a different thing, causing the children to unmount and mount again, losing its focus state. So instead, always pass in `this._focus` as `onClick` but do nothing if we're already focused. This comment should probably be near that `onClick`?

Reply:
yeah that would be better. Maybe just say `keep a constant prop for onClick to avoid render thrashing`